doc/xgboost_on_sqlflow.md

```diff
@@ -0,0 +1,139 @@
+
+XGBoost (https://xgboost.ai/) is an optimized distributed gradient boosting library designed to be highly efficient, 
+flexible and portable, which is often regarded as one of the best GBM frameworks.
+
+
+## xgboost on sqlflow via ant-xgboost
+   
+### Overview
+
+We use [ant-xgboost](https://github.com/alipay/ant-xgboost) as backend,
+which is forked from [xgboost](https://github.com/dmlc/xgboost).
+_Ant-xgboost_ is nearly same as original xgboost, 
+except some improvements to make xgboost easier to use, 
+such as better early stopping strategy, parameter checking, and end to end launcher.
+
+### User Experience
+    
+In terms of sqlflow users, _xgboost_ is an alternative _Estimator_ like _Tensorflow Estimators_. 
+Working with xgboost is quite similar to working with Tensorflow estimators; 
+just change `TRAIN DNNClassifier` into `TRAIN XGBoostEstimator`. 
+In addition, some xgboost specific parameters are required, 
+which can be configured in the same way as _Tensorflow_ parameters. 
+
+Below is a demo about training/predicting via xgboost :
+
+```sql
+// sample clause of train
+select 
+    c1, c2, c3, c4, c5 as class
+from kaggle_credit_fraud_training_data
+TRAIN XGBoostEstimator
+WITH
+  booster = "gbtree"
+  objective = "logistic:binary"
+  eval_metric = "auc"
+  train_eval_ratio = 0.8
+COLUMN
+  NUMERIC(c1) as fc1,
+  BUCKETIZED(fc1, [0, 10, 100]) as fc2,
+  CROSS([fc1, fc2, c3]) as fc3,
+  c4 as fc4
+LABEL class
+INTO sqlflow_models.xgboost_model_table;
+
+// sample clause of predict
+select 
+    c1, c2, c3, c4
+from kaggle_credit_fraud_development_data
+PREDICT kaggle_credit_fraud_development_data.class
+USING sqlflow_models.xgboost_model_table
+COLUMN
+  NUMERIC(c1) as fc1,
+  BUCKETIZED(fc1, [0, 10, 100]) as fc2,
+  CROSS([fc1, fc2, c3]) as fc3,
+  c4 as fc4;
+```
+
+### Implementation
+
+As `codegen.go` generating _TensorFlow_ code from sqlflow AST,
+we will add `codegen_xgboost.go` which translate sqlflow AST into a python launcher program of _xgboost_. 
+
+In _ant-xgboost_, there exists an incubating module named [_xgblauncher_](https://github.com/alipay/ant-xgboost/tree/ant_master/xgboost-launcher), 
+an extendable, cloud-native xgboost based machine learning pipeline, 
+which provides a python API for building custom `DataSource` and `ModelSource`.
+
+Below is a demonstration of DataSource/ModelSource API.
+ 
+```python
+class DataSource:
+    """
+    DataSource API
+    A handler of data reading/writing, which is compatible with both single-machine and distributed runtime.
+    """
+    def __init__(self, 
+                 rank: int, 
+                 num_worker: int,
+                 column_conf: configs.ColumnFields,
+                 source_conf):
+        pass
+        
+    @abstractmethod
+    def read(self) -> Iterator[XGBoostRecord]:
+        pass
+
+    @abstractmethod
+    def write(self, result_iter: Iterator[XGBoostResult]):
+        pass
+
+    
+class ModelSource:
```

Comment:
How is this class different from XgBoost's `save/load`? [ref](https://xgboost.readthedocs.io/en/latest/python/python_intro.html#training)

Reply:
`ModelSource`is a wrapper of `booster.save/load`,  it integrate saving work of booster binary, booster info and  feature scores.  The purpose of `MdoelSource` is prividing an interface to save/load xgboost models on differenct stores. 
Here is the full implemention:  https://github.com/alipay/ant-xgboost/blob/ant_master/xgboost-launcher/launcher/model_source.py